shared/src/main/scala/datasource.scala

```diff
@@ -25,21 +25,26 @@ import cats.syntax.all._
 
 /**
  * A `DataSource` is the recipe for fetching a certain identity `I`, which yields
- * results of type `A`.
+ * results of type `A` performing an effect of type `F[_]`.
  */
-trait DataSource[I, A] {
+trait DataSource[F[_], I, A] {
   /** The name of the data source.
    */
   def name: String
 
+  /** The identity of the data source.
+   */
+  def identity: DataSourceIdentity =
+    DataSourceIdentity(name)
+
   /** Fetch one identity, returning a None if it wasn't found.
    */
-  def fetch[F[_] : ConcurrentEffect](id: I): F[Option[A]]
+  def fetch(id: I)(implicit C: ConcurrentEffect[F]): F[Option[A]]
```

Comment:
Do you need this implicit parameter on the method itself, or can it be delegated as a constructor parameter of the implementing classes?

If you leave it for the implementation, you can still have a tests mock classes where you do not need the restrictions, just the `Applicative` to use a `pure`.

Reply:
> Do you need this implicit parameter on the method itself, or can it be delegated as a constructor parameter of the implementing classes?

Yes, I want it to be available in the methods to use the `ConcurrentEffect` API and have that implicit in scope.

Reply:
Right, but the benefit of an `implicit` value, or type-class instance, is that you do not need to feed it in the super-class. You can let a particular subclass to require it when it uses it in the implementation. IMHO, we should lean on using the _abstract_ methods to just declare a dependency of input parameters (the `id`) to the output (the `Option[A]`).